### PR TITLE
Backports

### PR DESCRIPTION
**Pull request by jhunkeler**
CC @cslocum

All OOB commits from SVN

SVN checkout via:
```git svn clone https://svn.stsci.edu/svn/ssb/etal/pandokia/trunk```

Patch set generated with:
```git format-patch --diff-algorithm=minimal a8b2d71fb2a96b5703753c66621c311dd26f0420```

Patch set applied with:
```git am -3 /path/to/*.patch```

### CODE REVIEW

**Comment by jhunkeler** [comment]
# Incremental patchset

```diff
From c00fcbfa3a02a987d4472046e0eb29e6f9f1eee4 Mon Sep 17 00:00:00 2001
From: cslocum <cslocum@d34015c8-bcbb-4646-8ac8-8ba5febf221d>
Date: Fri, 3 Mar 2017 19:56:37 +0000
Subject: [PATCH 1/7] Add transaction-level debugging

git-svn-id: https://svn.stsci.edu/svn/ssb/etal/pandokia/trunk@1548 d34015c8-bcbb-4646-8ac8-8ba5febf221d
---
 pandokia/db_mysqldb.py | 31 +++++++++++++++++++++----------
 1 file changed, 21 insertions(+), 10 deletions(-)

diff --git a/pandokia/db_mysqldb.py b/pandokia/db_mysqldb.py
index 2205f9d..9475623 100644
--- a/pandokia/db_mysqldb.py
+++ b/pandokia/db_mysqldb.py
@@ -20,8 +20,7 @@ thread_safe = db_module.threadsafety
 import pandokia.db
 
 # debugging
-_tty = None
-# _tty = open("/dev/tty","w")
+debug = False
 
 try:
     import io as StringIO
@@ -163,15 +162,27 @@ class PandokiaDB(pandokia.db.where_dict_base):
         # create a cursor, execute the statement
         c = self.db.cursor()
 
-        if (_tty is not None) and not (statement.startswith('EXPLAIN')):
-            _tty.write(
-                "--------\nQUERY: %s\nparam %s\n" %
-                (statement, str(parameters)))
-            _tty.write(explain_query(statement, parameters) + "\nWARN: ")
+        if debug:
+            stmt_keys_start = statement.find('(')
+            stmt_keys_stop = statement.find(')')
+            stmt_keys = [ x for x in statement[stmt_keys_start + 1: stmt_keys_stop - 1].replace(',', '').strip().split()]
+            pmtr_keys = [ int(x) for x in sorted(parameters.keys(), key=lambda i: int(i))]
+            keymap = zip(pmtr_keys, stmt_keys)
+
+            print('QUERY: {0}'.format(statement))
+            print('PARAM:')
+            for k, v in sorted(parameters.items(), key=lambda i: int(i[0])):
+                for keys in keymap:
+                    if int(k) == keys[0]:
+                        k = keys[1]
+                        break
+
+                k = str(k)
+                v = str(v)
+                print('{0:<16s} ({1:>6d}) = {2}'.format(k, len(v), v))
+
             c.execute("SHOW WARNINGS")
-            for x in c:
-                _tty.write(str(x) + "\n")
-            _tty.write("\n\n\n")
+            print("\n\n\n")
 
         # print parameters,"<br>"
         c.execute(statement, parameters)
-- 
2.13.2


From 60d03a3dd68cd74921d1bc3015e1c7bbb33f75cf Mon Sep 17 00:00:00 2001
From: cslocum <cslocum@d34015c8-bcbb-4646-8ac8-8ba5febf221d>
Date: Fri, 24 Feb 2017 15:11:38 +0000
Subject: [PATCH 2/7] increment number in go to 92

git-svn-id: https://svn.stsci.edu/svn/ssb/etal/pandokia/trunk@1547 d34015c8-bcbb-4646-8ac8-8ba5febf221d
---
 go | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/go b/go
index 3c652ad..b8e187b 100755
--- a/go
+++ b/go
@@ -2,7 +2,7 @@
 # This is a script I use to install pandokia at various places that I use here at stsci.  This is unlikely to be useful to you.
 echo $path
 
-set n=91
+set n=92
 
 rm -rf build
 
-- 
2.13.2


From edbd4b0dae010077693c096bcd6c3a603eef0a99 Mon Sep 17 00:00:00 2001
From: cslocum <cslocum@d34015c8-bcbb-4646-8ac8-8ba5febf221d>
Date: Fri, 3 Mar 2017 20:02:27 +0000
Subject: [PATCH 3/7] Increase test_run,test_name,project,context limits

git-svn-id: https://svn.stsci.edu/svn/ssb/etal/pandokia/trunk@1549 d34015c8-bcbb-4646-8ac8-8ba5febf221d
---
 pandokia/sql/mysql.sql | 22 +++++++++++-----------
 1 file changed, 11 insertions(+), 11 deletions(-)

diff --git a/pandokia/sql/mysql.sql b/pandokia/sql/mysql.sql
index dcd3988..e02e664 100644
--- a/pandokia/sql/mysql.sql
+++ b/pandokia/sql/mysql.sql
@@ -44,9 +44,9 @@ CREATE TABLE result_scalar (
 			PRIMARY KEY ( key_id ),
 		-- primary key is assigned by database; this is a
 		-- unique identifier
-	test_run 	VARCHAR(50),
-	project 	VARCHAR(25),
-	context		VARCHAR(25),
+	test_run 	VARCHAR(200),
+	project 	VARCHAR(200),
+	context		VARCHAR(200),
 	test_name 	VARCHAR(500),
 		-- this is the tuple that identifies a specific test
 		-- results
@@ -159,7 +159,7 @@ CREATE INDEX result_log_index
 --	convert a project/test_run/host to a list of email addresses
 
 CREATE TABLE contact (
-	project	VARCHAR(25),
+	project	VARCHAR(200),
 	test_name VARCHAR(500),
 	email VARCHAR(128)
 	);
@@ -177,10 +177,10 @@ CREATE TABLE expected (
 		-- this "daily_" or something like that; the information
 		-- that connects the test_run_type to an actual test_run
 		-- comes from outside the database.
-	project VARCHAR(25),
+	project VARCHAR(200),
 	host VARCHAR(64),
 	test_name VARCHAR(500),
-	context VARCHAR(25)
+	context VARCHAR(200)
 		-- project, host, test_name, context as in result_scalar
 	);
 
@@ -194,7 +194,7 @@ CREATE UNIQUE INDEX expected_unique
 --	make a table that just contains the distinct values.
 
 CREATE TABLE distinct_test_run (
-	test_run VARCHAR(50) UNIQUE,
+	test_run VARCHAR(200) UNIQUE,
 	valuable CHAR(1),
 		-- boolean, but portable; use 1 or 0
 		-- valuable means that we should not refuse to delete
@@ -224,7 +224,7 @@ CREATE UNIQUE INDEX user_prefs_username_index
 
 CREATE TABLE user_email_pref (
 	username VARCHAR(30),
-	project VARCHAR(25),
+	project VARCHAR(200),
 	format CHAR(1),
 		-- format is one of:
 		-- 'n' = none; send no email about this project
@@ -301,14 +301,14 @@ CREATE INDEX hostinfo_index
 --      chronic or not.
 
 CREATE TABLE chronic (
-	test_run_type VARCHAR(100),
+	test_run_type VARCHAR(200),
 		-- this "daily_" or something like that; the information
 		-- that connects the test_run_type to an actual test_run
 		-- comes from outside the database.
-	project VARCHAR(25),
+	project VARCHAR(200),
 	host VARCHAR(64),
 	test_name VARCHAR(500),
-	context VARCHAR(25),
+	context VARCHAR(200),
 		-- project, host, test_name, context as in result_scalar
         xwhen VARCHAR(50)
                 -- indicator of when the test first went bad
-- 
2.13.2


From 4ab3eab2cfae1cd3ba47160c92ef5d6c4a58134d Mon Sep 17 00:00:00 2001
From: cslocum <cslocum@d34015c8-bcbb-4646-8ac8-8ba5febf221d>
Date: Fri, 3 Mar 2017 20:07:13 +0000
Subject: [PATCH 4/7] Normalize sqlite tables; they didn't match mysql.sql

git-svn-id: https://svn.stsci.edu/svn/ssb/etal/pandokia/trunk@1550 d34015c8-bcbb-4646-8ac8-8ba5febf221d
---
 pandokia/sql/sqlite.sql | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/pandokia/sql/sqlite.sql b/pandokia/sql/sqlite.sql
index bd6cae4..2ae1860 100644
--- a/pandokia/sql/sqlite.sql
+++ b/pandokia/sql/sqlite.sql
@@ -173,7 +173,7 @@ CREATE UNIQUE INDEX expected_unique
 --	make a table that just contains the distinct values.
 
 CREATE TABLE distinct_test_run (
-	test_run VARCHAR(50) UNIQUE,
+	test_run VARCHAR(200) UNIQUE,
 	valuable CHAR(1),
 		-- boolean, but portable; use 1 or 0
 		-- valuable means that we should not refuse to delete
-- 
2.13.2


From 8cecd056a0716aaebf83de8191860302f30f6150 Mon Sep 17 00:00:00 2001
From: cslocum <cslocum@d34015c8-bcbb-4646-8ac8-8ba5febf221d>
Date: Fri, 3 Mar 2017 20:15:58 +0000
Subject: [PATCH 5/7] incremenr n to 93

git-svn-id: https://svn.stsci.edu/svn/ssb/etal/pandokia/trunk@1551 d34015c8-bcbb-4646-8ac8-8ba5febf221d
---
 go | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/go b/go
index b8e187b..2ac417c 100755
--- a/go
+++ b/go
@@ -2,7 +2,7 @@
 # This is a script I use to install pandokia at various places that I use here at stsci.  This is unlikely to be useful to you.
 echo $path
 
-set n=92
+set n=93
 
 rm -rf build
 
-- 
2.13.2


From 58d44eb5b9515a90ae25f7a0880869268cd7c5a0 Mon Sep 17 00:00:00 2001
From: cslocum <cslocum@d34015c8-bcbb-4646-8ac8-8ba5febf221d>
Date: Fri, 24 Mar 2017 12:34:14 +0000
Subject: [PATCH 6/7] Added recurring prefix 'pandeia_master_latest', bumped N
 value for testing

git-svn-id: https://svn.stsci.edu/svn/ssb/etal/pandokia/trunk@1554 d34015c8-bcbb-4646-8ac8-8ba5febf221d
---
 go                         | 2 +-
 pandokia/default_config.py | 1 +
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/go b/go
index 2ac417c..cbef41c 100755
--- a/go
+++ b/go
@@ -2,7 +2,7 @@
 # This is a script I use to install pandokia at various places that I use here at stsci.  This is unlikely to be useful to you.
 echo $path
 
-set n=93
+set n=94
 
 rm -rf build
 
diff --git a/pandokia/default_config.py b/pandokia/default_config.py
index 18a2b9b..a4f6b81 100644
--- a/pandokia/default_config.py
+++ b/pandokia/default_config.py
@@ -254,6 +254,7 @@ recurring_prefix = (
     'etc_jwst_daily',
     'pandeia_nightly',
     'jwst',
+    'pandeia_master_latest'
 )
 
 #####
-- 
2.13.2


From d734c17e465187a8d10a4db61b50e4bba4c8ac4a Mon Sep 17 00:00:00 2001
From: cslocum <cslocum@d34015c8-bcbb-4646-8ac8-8ba5febf221d>
Date: Fri, 7 Apr 2017 18:22:36 +0000
Subject: [PATCH 7/7] switch from stat to os.isdir to allow symlink traversal

git-svn-id: https://svn.stsci.edu/svn/ssb/etal/pandokia/trunk@1558 d34015c8-bcbb-4646-8ac8-8ba5febf221d
---
 pandokia/run_recursive.py | 11 +----------
 1 file changed, 1 insertion(+), 10 deletions(-)

diff --git a/pandokia/run_recursive.py b/pandokia/run_recursive.py
index 644b368..b882751 100644
--- a/pandokia/run_recursive.py
+++ b/pandokia/run_recursive.py
@@ -43,16 +43,7 @@ def generate_directories(dir):
 
         # Find out if it is a directory.  If not, skip it.
         full_name = os.path.join(dir,short_name)
-        try:
-            # lstat - not recursing into symlinks
-            file_stat = os.lstat(full_name)
-        except OSError as e:
-            if e.errno != errno.ENOENT:
-                print("cannot lstat %s" % full_name)
-                print(e)
-            continue
-
-        if not stat.S_ISDIR(file_stat.st_mode):
+        if not os.path.isdir(full_name):
             continue
 
         # It is a directory - recursively search it.
-- 
2.13.2


```

**Comment by jhunkeler** [comment]
# Full diff
```diff
diff --git a/go b/go
index 3c652ad..cbef41c 100755
--- a/go
+++ b/go
@@ -2,7 +2,7 @@
 # This is a script I use to install pandokia at various places that I use here at stsci.  This is unlikely to be useful to you.
 echo $path
 
-set n=91
+set n=94
 
 rm -rf build
 
diff --git a/pandokia/db_mysqldb.py b/pandokia/db_mysqldb.py
index 2205f9d..9475623 100644
--- a/pandokia/db_mysqldb.py
+++ b/pandokia/db_mysqldb.py
@@ -20,8 +20,7 @@ thread_safe = db_module.threadsafety
 import pandokia.db
 
 # debugging
-_tty = None
-# _tty = open("/dev/tty","w")
+debug = False
 
 try:
     import io as StringIO
@@ -163,15 +162,27 @@ class PandokiaDB(pandokia.db.where_dict_base):
         # create a cursor, execute the statement
         c = self.db.cursor()
 
-        if (_tty is not None) and not (statement.startswith('EXPLAIN')):
-            _tty.write(
-                "--------\nQUERY: %s\nparam %s\n" %
-                (statement, str(parameters)))
-            _tty.write(explain_query(statement, parameters) + "\nWARN: ")
+        if debug:
+            stmt_keys_start = statement.find('(')
+            stmt_keys_stop = statement.find(')')
+            stmt_keys = [ x for x in statement[stmt_keys_start + 1: stmt_keys_stop - 1].replace(',', '').strip().split()]
+            pmtr_keys = [ int(x) for x in sorted(parameters.keys(), key=lambda i: int(i))]
+            keymap = zip(pmtr_keys, stmt_keys)
+
+            print('QUERY: {0}'.format(statement))
+            print('PARAM:')
+            for k, v in sorted(parameters.items(), key=lambda i: int(i[0])):
+                for keys in keymap:
+                    if int(k) == keys[0]:
+                        k = keys[1]
+                        break
+
+                k = str(k)
+                v = str(v)
+                print('{0:<16s} ({1:>6d}) = {2}'.format(k, len(v), v))
+
             c.execute("SHOW WARNINGS")
-            for x in c:
-                _tty.write(str(x) + "\n")
-            _tty.write("\n\n\n")
+            print("\n\n\n")
 
         # print parameters,"<br>"
         c.execute(statement, parameters)
diff --git a/pandokia/default_config.py b/pandokia/default_config.py
index 18a2b9b..a4f6b81 100644
--- a/pandokia/default_config.py
+++ b/pandokia/default_config.py
@@ -254,6 +254,7 @@ recurring_prefix = (
     'etc_jwst_daily',
     'pandeia_nightly',
     'jwst',
+    'pandeia_master_latest'
 )
 
 #####
diff --git a/pandokia/run_recursive.py b/pandokia/run_recursive.py
index 644b368..b882751 100644
--- a/pandokia/run_recursive.py
+++ b/pandokia/run_recursive.py
@@ -43,16 +43,7 @@ def generate_directories(dir):
 
         # Find out if it is a directory.  If not, skip it.
         full_name = os.path.join(dir,short_name)
-        try:
-            # lstat - not recursing into symlinks
-            file_stat = os.lstat(full_name)
-        except OSError as e:
-            if e.errno != errno.ENOENT:
-                print("cannot lstat %s" % full_name)
-                print(e)
-            continue
-
-        if not stat.S_ISDIR(file_stat.st_mode):
+        if not os.path.isdir(full_name):
             continue
 
         # It is a directory - recursively search it.
diff --git a/pandokia/sql/mysql.sql b/pandokia/sql/mysql.sql
index dcd3988..e02e664 100644
--- a/pandokia/sql/mysql.sql
+++ b/pandokia/sql/mysql.sql
@@ -44,9 +44,9 @@ CREATE TABLE result_scalar (
 			PRIMARY KEY ( key_id ),
 		-- primary key is assigned by database; this is a
 		-- unique identifier
-	test_run 	VARCHAR(50),
-	project 	VARCHAR(25),
-	context		VARCHAR(25),
+	test_run 	VARCHAR(200),
+	project 	VARCHAR(200),
+	context		VARCHAR(200),
 	test_name 	VARCHAR(500),
 		-- this is the tuple that identifies a specific test
 		-- results
@@ -159,7 +159,7 @@ CREATE INDEX result_log_index
 --	convert a project/test_run/host to a list of email addresses
 
 CREATE TABLE contact (
-	project	VARCHAR(25),
+	project	VARCHAR(200),
 	test_name VARCHAR(500),
 	email VARCHAR(128)
 	);
@@ -177,10 +177,10 @@ CREATE TABLE expected (
 		-- this "daily_" or something like that; the information
 		-- that connects the test_run_type to an actual test_run
 		-- comes from outside the database.
-	project VARCHAR(25),
+	project VARCHAR(200),
 	host VARCHAR(64),
 	test_name VARCHAR(500),
-	context VARCHAR(25)
+	context VARCHAR(200)
 		-- project, host, test_name, context as in result_scalar
 	);
 
@@ -194,7 +194,7 @@ CREATE UNIQUE INDEX expected_unique
 --	make a table that just contains the distinct values.
 
 CREATE TABLE distinct_test_run (
-	test_run VARCHAR(50) UNIQUE,
+	test_run VARCHAR(200) UNIQUE,
 	valuable CHAR(1),
 		-- boolean, but portable; use 1 or 0
 		-- valuable means that we should not refuse to delete
@@ -224,7 +224,7 @@ CREATE UNIQUE INDEX user_prefs_username_index
 
 CREATE TABLE user_email_pref (
 	username VARCHAR(30),
-	project VARCHAR(25),
+	project VARCHAR(200),
 	format CHAR(1),
 		-- format is one of:
 		-- 'n' = none; send no email about this project
@@ -301,14 +301,14 @@ CREATE INDEX hostinfo_index
 --      chronic or not.
 
 CREATE TABLE chronic (
-	test_run_type VARCHAR(100),
+	test_run_type VARCHAR(200),
 		-- this "daily_" or something like that; the information
 		-- that connects the test_run_type to an actual test_run
 		-- comes from outside the database.
-	project VARCHAR(25),
+	project VARCHAR(200),
 	host VARCHAR(64),
 	test_name VARCHAR(500),
-	context VARCHAR(25),
+	context VARCHAR(200),
 		-- project, host, test_name, context as in result_scalar
         xwhen VARCHAR(50)
                 -- indicator of when the test first went bad
diff --git a/pandokia/sql/sqlite.sql b/pandokia/sql/sqlite.sql
index bd6cae4..2ae1860 100644
--- a/pandokia/sql/sqlite.sql
+++ b/pandokia/sql/sqlite.sql
@@ -173,7 +173,7 @@ CREATE UNIQUE INDEX expected_unique
 --	make a table that just contains the distinct values.
 
 CREATE TABLE distinct_test_run (
-	test_run VARCHAR(50) UNIQUE,
+	test_run VARCHAR(200) UNIQUE,
 	valuable CHAR(1),
 		-- boolean, but portable; use 1 or 0
 		-- valuable means that we should not refuse to delete

```

**Comment by vglaidler** [comment]
I looked at the svn diffs & had some questions about a couple of those revisions I want to do forensics and/or talk to Christine about.

File: pandokia/default_config.py [hunk 1]
```diff
@@ -254,6 +254,7 @@ def complex_readpass():
     'etc_jwst_daily',
     'pandeia_nightly',
     'jwst',
+    'pandeia_master_latest'
```
Comment (vglaidler) [inline pandokia/default_config.py]: This is one of the things I want to doublecheck - "latest" is magic-substitution & I wouldn't expect it to appear in a pattern, & am not sure how it interacts with the magic.

**Pull request merged**